api/v2beta1/helmrelease_types.go

```diff
@@ -463,6 +476,24 @@ type Upgrade struct {
 	// upgrade action when it fails.
 	// +optional
 	CleanupOnFail bool `json:"cleanupOnFail,omitempty"`
+
+	// UpgradeCRDs upgrade CRDs from the Helm Chart's crds directory according
+	// to the CRD upgrade policy provided here. Valid values are `Create` or
+	// `CreateReplace`. If omitted (the default) CRDs
+	// are not upgraded.
+	//
+	// Create: new CRDs are created, existing CRDs are neither updated nor deleted.
+	//
+	// CreateReplace: new CRDs are created, existing CRDs are updated (replaced)
+	// but not deleted.
+	//
+	// By default, CRDs are not applied during Helm upgrade action. With this
+	// option users can opt-in to CRD upgrade, which is not (yet) natively supported by Helm.
+	// https://helm.sh/docs/chart_best_practices/custom_resource_definitions.
+	//
+	// +kubebuilder:validation:Enum=Create;CreateReplace
+	// +optional
+	UpgradeCRDs CRDsChangePolicy `json:"upgradeCRDs,omitempty"`
```

Comment:
Having looked at the implementation now, I see room for install to support using the `CreateReplace` policy during install, given all logic is there.

However, if we would do this, I think the flags probably should be changed to be included in the action specific configurations (`spec.install` and `spec.upgrade`).

```yaml
spec:
  # Fallback default
  crds: "None"
  install:
    crds: "Create"
  upgrade:
	# Default: "None"
    crds: "CreateReplace"
```

When `install.crds == Create` it is probably best to use the Helm config flag instead of our own logic.

The old configuration flag should be marked deprecated but still taken into account.

What do you think?

Reply:
I am not yet sure how to implement this, but let me share the key tenets that I have in mind:

- Keep backward compatibility, this PR should not introduce any breaking change to existing `HelmRelease` objects. This implies that for `install` the default must be `Create` and for `upgrade` the default must be `Skip`.
- Helm install should (whenever possible) rely on the functionality provided by Helm and only use our own logic where Helm has no native support. This is important to make sure HelmController's behavior stays as close to Helm's as possible. 

From above tenets we can see that `install` and `upgrade` have different default values (for backward compatibility resp. to provide Helm behavior by default). Therefore, I am inclined towards not providing a global `crds` attribute 
and to go with the action specific `crds` attributes instead.

Let's extend the policy enum like this:

`crds := Skip | Create | CreateReplace`

```
spec:
  install:
    # Default: "Create"
    crds: "Create"
    # Default false (we have to keep this for backward compatibility and might remove it in a future API version)
    skipCRDs: false
  upgrade:
    # Default: "Skip"
    crds: "CreateReplace"
```

And we can implement the following logic:

|action|`crds` value|Behaviour|
|---|---|---|
|`install`|`Skip`|Same as `install.skipCRDs = true`, pass skipCRDs flag to Helm|
|`install`|`Create`|Same as `install.skipCRDs = false`, let Helm create the CRDs for us|
|`install`|`CreateReplace`|Use our own logic, as Helm does not support this natively|
|`upgrade`|`Skip`|Do nothing at all|
|`upgrade`|`Create`|Use our own logic, as Helm does not support this natively|
|`upgrade`|`CreateReplace`|Use our own logic, as Helm does not support this natively|

What should happen if the the user specifies contradicting `crds` and `skipCRDs` on `install`?

```
spec:
  install:
    crds: "Create"
    skipCRDs: true
```

Shall we fail with an appropriate error message (my preference) or shall we defined a precedence rule?

What do you think?

Reply:
Sounds like a solid plan.

> What should happen if the the user specifies contradicting `crds` and `skipCRDs` on install?
> Shall we fail with an appropriate error message (my preference) or shall we defined a precedence rule?

Due to deprecation giving `crds` precedence over `skipCRDs` would be an OK thing to do. But returning an error will save some users but annoy others, and usually saving a minority at the cost of annoying another is the right thing to do.

Reply:
OK, I implemented this, ready for review.